zebra-chain/src/orchard/sinsemilla.rs

```diff
@@ -0,0 +1,150 @@
+pub fn extract_p(point: pallas::Point) -> pallas::Base {
+    match pallas::Affine::from(point).get_xy().into() {
+        // If Some, it's not the identity.
+        Some((x, _)) => x,
+        _ => pallas::Base::zero(),
+    }
+}
+
+/// GroupHash into Pallas, aka _GroupHash^P_
+///
+/// Produces a random point in the Pallas curve.  The first input element acts
+/// as a domain separator to distinguish uses of the group hash for different
+/// purposes; the second input element is the message.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretegrouphashpallasandvesta
+#[allow(non_snake_case)]
+pub fn pallas_group_hash(D: &[u8], M: &[u8]) -> pallas::Point {
+    let domain_separator = std::str::from_utf8(D).unwrap();
+
+    pallas::Point::hash_to_curve(domain_separator)(M)
+}
+
+/// Q(D) := GroupHash^P(︀“z.cash:SinsemillaQ”, D)
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+fn Q(D: &[u8]) -> pallas::Point {
+    pallas_group_hash(b"z.cash:SinsemillaQ", D)
+}
+
+/// S(j) := GroupHash^P(︀“z.cash:SinsemillaS”, LEBS2OSP32(I2LEBSP32(j)))
+///
+/// S: {0 .. 2^k - 1} -> P^*, aka 10 bits hashed into the group
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+#[allow(non_snake_case)]
+fn S(j: &u16) -> pallas::Point {
+    // The value of j is a 10-bit value, therefore must never exceed 2^10 in
+    // value.
+    assert!(j < &1024u16);
+
+    pallas_group_hash(b"z.cash:SinsemillaS", &j.to_le_bytes())
+}
+
+/// "...an algebraic hash function with collision resistance (for fixed input
+/// length) derived from assumed hardness of the Discrete Logarithm Problem on
+/// the Pallas curve."
+///
+/// SinsemillaHash is used in the definitions of Sinsemilla commitments and of
+/// the Sinsemilla hash for the Orchard incremental Merkle tree (§ 5.4.1.3
+/// ‘MerkleCRH^Orchard Hash Function’).
+///
+/// https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+///
+/// # Panics
+///
+/// If `M` is greater than `k*c = 2530` bits.
+#[allow(non_snake_case)]
+pub fn sinsemilla_hash_to_point(D: &[u8], M: &BitVec<Lsb0, u8>) -> pallas::Point {
+    let k = 10;
+    let c = 253;
+
+    assert!(M.len() <= k * c);
+
+    let mut acc = Q(D);
+
+    // Split M into n segments of k bits, where k = 10 and c = 253, padding
+    // the last segment with zeros.
+    //
+    // https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+    for chunk in M.chunks(k) {
+        // Pad each chunk with zeros.
+        let mut store = 0u16;
+        let bits = store.bits_mut::<Lsb0>();
+        chunk
+            .iter()
+            .enumerate()
+            .for_each(|(i, bit)| bits.set(i, *bit));
+
+        // An instance of LEBS2IP_k
+        // XXX: does Rust or bitvec have a better implementation?
+        let j = &bits
+            .iter()
+            .enumerate()
+            .fold(0u16, |j, (i, &bit)| j + if bit { 1 << i } else { 0 });
+
+        acc = acc + acc + S(j);
```

Comment:
This looks a lot like `copy_from_bitslice`:
https://docs.rs/bitvec/0.22.3/bitvec/slice/struct.BitSlice.html#method.copy_from_bitslice

```suggestion
        // An instance of LEBS2IP_k
        let bits = store.bits_mut::<Lsb0>()[..k];
        bits.copy_from_bitslice(chunk);
        
        acc = acc + acc + S(store);
```

Reply:
`acc + acc + S(store)` isn't equivalent to `(acc ⸭ S(store)) ⸭ acc` where `⸭` is incomplete addition.

Reply:
> `acc + acc + S(store)` isn't equivalent to `(acc ⸭ S(store)) ⸭ acc` where `⸭` is checked incomplete addition.

I'll double check on this in [the follow-on PR ](https://github.com/ZcashFoundation/zebra/issues/2079)where I pull in the test vectors from `zcash-hackworks` to exercise this 👍

Reply:
> This looks a lot like `copy_from_bitslice`:
> https://docs.rs/bitvec/0.22.3/bitvec/slice/struct.BitSlice.html#method.copy_from_bitslice

I used this in part in 42f7bb5c, since we're on 0.17.4 I had to change it a bit.